Heat equation smoothing of IC

wilfonba opened the pull request:
## Description

This PR adds heat equation smoothing for ICs.

### Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

### Scope

- [x] This PR comprises a set of related changes with a common goal

## How Has This Been Tested?

The following figures show a bunch of smoothing configurations, with black borders denoting processor boundaries:

The following examples were ran on Wingtip with 64 CPU ranks:

<img width="424" alt="image" src="https://github.com/user-attachments/assets/b5fe7c6d-7f88-4cf7-9cd2-b97911147f7f" />
<img width="424" alt="image" src="https://github.com/user-attachments/assets/f08ac64e-ba89-4e2e-8699-3ea4897a09bb" />
<img width="424" alt="image" src="https://github.com/user-attachments/assets/43630df3-32d7-48a3-8ffb-1c2c371209ba" />
<img width="424" alt="image" src="https://github.com/user-attachments/assets/07d9e5d8-24ea-4131-a77a-20ba58d05b83" />
<img width="498" alt="image" src="https://github.com/user-attachments/assets/9ddf66be-341b-470e-8fc5-a87ee9186fc6" />
<img width="498" alt="image" src="https://github.com/user-attachments/assets/61482af4-6608-4128-b385-e696dd8dd421" />
<img width="498" alt="image" src="https://github.com/user-attachments/assets/720a4f74-e1ee-4009-9ddd-07aa0be2e558" />
<img width="498" alt="image" src="https://github.com/user-attachments/assets/e8e4a3e8-daa2-4daa-b5f2-d2dbe6ca48c8" />

## Checklist

- [x] I have added comments for the new code
- [x] I have made corresponding changes to the documentation (`docs/`)
- [x] I have added regression tests to the test suite so that people can verify in the future that the feature is behaving as expected
- [x] I ran `./mfc.sh format` before committing my code
- [x] New and existing tests pass locally with my changes, including with GPU capability enabled (both NVIDIA hardware with NVHPC compilers and AMD hardware with CRAY compilers) and disabled
- [x] This PR does not introduce any repeated code (it follows the [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) principle)
- [x] I cannot think of a way to condense this code and reduce any introduced additional line count

### If your code changes any code source files (anything in `src/simulation`)

To make sure the code is performing as expected on GPU devices, I have:
- [x] Checked that the code compiles using NVHPC compilers
- [x] Checked that the code compiles using CRAY compilers
- [x] Ran the code on either V100, A100, or H100 GPUs and ensured the new feature performed as expected (the GPU results match the CPU results)
- [x] Ran the code on MI200+ GPUs and ensure the new features performed as expected (the GPU results match the CPU results)
- [x] Ran my code using various numbers of different GPUs (1, 2, and 8, for example) in parallel and made sure that the results scale similarly to what happens if you run without the new code/feature

The following shows strong and weak scaling on Frontier with and without RDMA mpi. Strong scaling was done with a 64m grid cell `3D_performance_test` and starts at two ranks and ends at 32 ranks. Weak scaling was done with 8m grid cells per GPU using the same case file, beginning at two ranks and ending at 32 ranks.
![image](https://github.com/user-attachments/assets/7e2a1e2b-9b60-422e-b968-0678314d3a1b)

sbryngelson reviewed (changes requested):
+700 lines for just this one feature? is there anyway to trim this down?

sbryngelson commented on src/pre_process/m_mpi_proxy.fpp:
```diff
@@ -538,4 +635,206 @@ contains
+        !!  @param q_prim_vf Cell-average primitive variables
+        !!  @param mpi_dir MPI communication coordinate direction
+        !!  @param pbc_loc Processor boundary condition (PBC) location
+    subroutine s_mpi_sendrecv_variables_buffers(q_prim_vf, &
```
isn't this subroutine repeated from simulation?

sbryngelson commented on src/pre_process/m_global_parameters.fpp:
```diff
@@ -779,12 +797,32 @@ contains
+        ! the physical computational domain from one time-step iteration to
+        ! the next one
+        if (viscous) then
+            buff_size = 2*weno_polyn + 2
```
this is duplicate code from simulation. move it somewhere common, along with all duplicate code

sbryngelson commented on src/pre_process/m_perturbation.fpp:
```diff
@@ -606,4 +616,290 @@ contains
+                        end do
+                    end do
+                end do
+            else if (bcxb == -2) then
```
this can't be the best way to do this...

wilfonba replied:
Addressed

wilfonba commented:
@sbryngelson, I've addressed your comments. Now that I've modified code in `simulation,` I'll have to do more extensive testing tomorrow before this can be merged.
